File: graphColoringAlgo/greedyGraphColoring/greedyGraphColoring.java
import  java.util.*;

class greedyGraphColoring{

    //function to return adjaceny list for graph for specified number of vertices
    static ArrayList<ArrayList<Integer>> graphCreate(int v){
        ArrayList<ArrayList<Integer> > graph= new ArrayList<ArrayList<Integer> >(v);
        for (int i=0;i<v;i++){
            graph.add(new ArrayList<Integer>());
        }
        return graph;
    }

    //function to create adjacency list representation of graph
    static void addEdge( ArrayList<ArrayList<Integer>> graph , int a , int b){
        graph.get(a).add(b);
        graph.get(b).add(a);
    }

    //greedy algorithm to solve graph coloring problem
    static String[] greedyGraphColoring(ArrayList<ArrayList<Integer>> graph,String[] colors){

        int n= graph.size();
        int color=colors.length;

        //Creating a list vertices to assign colours to
        int[] vertices=new int[graph.size()];
        for(int i=0;i<n;i++){
            vertices[i]=-1;
        }
    
        
        for(int ii=0;ii<n;ii++){


            ArrayList<Integer> neighbours=graph.get(ii);
            ArrayList<Integer> forbidden= new ArrayList<Integer>();
            for(int i=0;i<neighbours.size();i++){
                if(vertices[neighbours.get(i)]!=-1){
                    forbidden.add(vertices[neighbours.get(i)]);
                }
            }

            //Assigning color to vertice 
            for(int i=0;i<color;i++){
                if(!forbidden.contains(i)){
                    vertices[ii]=i;
                    break;
                }
            }

        

        }

        //Mapping the colors
        String[] verticeColors= new String[n];
        for(int i=0;i<n;i++){
            verticeColors[i]= colors[vertices[i]];
        }
        return verticeColors;
    }

    
    
    public static void main(String[] args)
    {
        // Creating a graph with 5 vertices
        int vertices = 5;
        ArrayList<ArrayList<Integer>> adj = graphCreate(vertices);
        String [] colors =new String[] {"red", "chocolate", "darkorange", "moccasin", "gold"};

  
        // Adding edges one by one.
        addEdge(adj, 0, 1);
        addEdge(adj, 0, 4);
        addEdge(adj, 1, 2);
        addEdge(adj, 1, 3);
        addEdge(adj, 1, 4);
        addEdge(adj, 2, 3);
        addEdge(adj, 3, 4);
       
        String [] verticeColors= greedyGraphColoring(adj,colors);

        // Printing vertices and their respective colors

        System.out.println("Node : Color");
        for(int i=0;i<adj.size();i++){
            String s= i+ ":" +verticeColors[i];
            System.out.println(s);
        }


    }
}

File: graphColoringAlgo/greedyGraphColoring/greedyGraphColoring.cpp
#include <bits/stdc++.h>
using namespace std;

class Graph
{
    int V;
    list<int> *adj;

public:
    Graph(int V)
    {
        this->V = V;
        adj = new list<int>[V];
    }

    void addEdge(int v, int w)
    {
        adj[v].push_back(w);
        adj[w].push_back(v);
    }

    void greedyColoring()
    {
        int result[V];
        result[0] = 0;
        for (int u = 1; u < V; u++)
            result[u] = -1;

        bool available[V];
        for (int cr = 0; cr < V; cr++)
            available[cr] = false;

        for (int u = 1; u < V; u++)
        {
            list<int>::iterator i;
            for (i = adj[u].begin(); i != adj[u].end(); ++i)
                if (result[*i] != -1)
                    available[result[*i]] = true;

            int cr;
            for (cr = 0; cr < V; cr++)
                if (available[cr] == false)
                    break;

            result[u] = cr;

            for (i = adj[u].begin(); i != adj[u].end(); ++i)
                if (result[*i] != -1)
                    available[result[*i]] = false;
        }

        for (int u = 0; u < V; u++)
            cout << "Vertex " << u << " ---> Color "
                 << result[u] << endl;
    }
};

int main()
{
    Graph g1(7);
    g1.addEdge(0, 4);
    g1.addEdge(0, 5);
    g1.addEdge(0, 6);
    g1.addEdge(1, 3);
    g1.addEdge(2, 3);
    g1.addEdge(3, 4);
    g1.addEdge(4, 5);
    g1.addEdge(5, 6);
    cout<<"Coloured graph : \n";
    g1.greedyColoring();

    return 0;
}
